New marketplace search component

mlaetitia opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Add a new search component to the new marketplace react page (under WooCommerce WP Admin).
This component is prepared to call the WooCommerce marketplace API when a search query is entered.

We still need to add locale and country to the query as Props when calling the component.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch and build wc admin on your local environment
2. Add the <Search> component to a page . I tested it by:
 - Adding `import Search from '../search/search';` to `plugins/woocommerce-admin/client/marketplace/components/content/content.tsx`
 - Add `return <Search>` instead of `return <Extensions>` on line 22
3. Open WP admin -> WooCommerce -> Marketplace and head to the place where you added the component. Confirm the new component is there and looks like the project design
4. IMPORTANT: Might be a good idea to add a div with a certain width around the `<Search>` component - I purposely did not set the width to it so it would adapt easily to different layouts.
5. Add a `debugger;`. line to `runSearch` in `plugins/woocommerce-admin/client/marketplace/components/search/search.tsx` to confirm that it's triggered both when clicking the magnifying glass icon and when hitting Enter on the keyboard.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

simranvijwani reviewed (commented):
I haven't tested the PR yet, but based on my observation of the code, I recommend using variables in CSS. 
Check out this [example CSS file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/products/fields/variations/variations.scss) from the woocommerce-admin plugin where we utilize variables. It seems like we have to use variables from the [Gutenberg base style](https://github.com/WordPress/gutenberg/tree/trunk/packages/base-styles). We have a stylesheet with available color variables [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.scss).

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
Just noting that the styling has changed quite a bit in the MVP design since this was opened for review.

mlaetitia replied:
True! the search component didn't, it seems, but I color codes to the variables.scss anyway :) Thanks!

kdevnel replied:
Nice one! We can finalise any other styling when we incorporate all the components together.

kdevnel reviewed (changes requested):
This seems to work as described in your testing. Just noting that the design and location of the component has changed quite a lot since you opened this for review so I've requested changes for that reason only.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
-	height: 50px;
+@import '../../stylesheets/_variables.scss';
+
+.marketplace-search-form{
```
```suggestion
.marketplace-search-form {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
+	background-color: $gutenberg-gray-100;
+	padding: 9px 5px;
+	display: flex;
+	.search-form__search-input{
```
```suggestion
	.search-form__search-input {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
+		font-size: 13px;
+		line-height: 20px;
+		
+		::placeholder{
```
```suggestion

		::placeholder {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
+			color: $gutenberg-gray-700;
+		}
+	}
+	input[type="text"]:focus{
```
```suggestion
	input[type='text']:focus {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
+		border: none;
+		box-shadow: none;
+	}
+	.search-form__search-button{
```
```suggestion
	.search-form__search-button {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search/search.scss:
```diff
@@ -1,5 +1,36 @@
+		background-color: transparent;
+		border: none;
+		border-radius: 0;
+		cursor:pointer;
```
```suggestion
		cursor: pointer;
```

kdevnel reviewed (approved):
Looking great, I'm going to approve this one. There are some linting errors in the automated checks. I've suggested fixes for them for now.

mlaetitia commented:
Thanks @kdevnel  - made all the changes to fix linters, only the changelog missing, but definitely agree we can do a big Feature Branch one at the end.
All is good - assigning this to you now!

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/search/search.tsx:
```diff
@@ -1,12 +1,94 @@
+				<input
+					id="search-query"
+					className="search-form__search-input"
+					type="text"
```
This should be of type "search" to help mobile devices and screen readers.

```suggestion
					type="search"
```

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/search/search.tsx:
```diff
@@ -1,12 +1,94 @@
+
+	const renderSearch = () => {
+		return (
+			<div className="marketplace-search-form">
```
_Probably_ not worth changing, but if this was a `<form>` instead of a `<div>`:

1. You'd get the `onSubmit` event for free (which would be automatically be triggered by clicking the `<button>` or by pressing enter in the `<input>`, so you wouldn't need to handle `onKeyUp` on the `<input>` (checking for enter and calling `runSearch`): you could just `runSearch` when the form submits.
2. You could set the action="" of the `<form>` to the WCCOM version of the search. This way, the search would work even if JavaScript failed. Alternatively, it could have an action="" that submitted to the same page and capture the request that way, but it's much more work!
3. `<form>`s act as a landmark for screen readers, so they can jump to it. (You can, and probably should, "fake" this by adding `role="form"` to your `<div>` if you're keeping it as a `<div>`.)

Again: this probably isn't worth changing, just wanted to put the note here for educational purposes!

kdevnel replied:
This is such great info as always! I don't know that I'd have realised all this myself. Definitely one to note for future!

Dan-Q reviewed (changes requested):
One minor a11y change suggested (`type="search"`).

One other comment (a11y/progressive) that I don't think's worth acting on but noting for educational purposes.

kdevnel commented:
@Dan-Q I've implemented your requested change now. Thanks for the review!

andfinally committed:
In-app marketplace search component. Cleaned up the commit history of this branch. Copied changes from these commits:

e9828422706176817e511778980005222aa36cc5
9ca2ae351c97fcd27ecd77a1464c2a9ca16de040
e47815705f3854bf50ff48d7975b7cf2f541614b
976811c458e67ae7fa107c8bf8554fdc3e809d85
46eafdf49fe39c12dee77d6ce0885bdeda527dea

Deleted unused import.
